### DISCUSSION TITLE
Handling multiple occurrences of the same tag in xml

### DISCUSSION THREAD

**Discussion by ebarzilay-godaddy** [mikefarah/yq#1792] [Q&A]
**Please describe your feature request.**

I wish I could use yq to handle xml inputs that *might* have multiple occurrences of the same tag.

(I'm not sure if an issue is the proper place or if there *is* some hack around it, feel free to point me elsewhere if so.)

The problem is that yq wraps multiple occurrences of a repeated tag as an array, but does *not* do that if there's a single occurrence.

A quick example of what this has lead me to -- given the following two files:

```
<foo>
  <bar x="xx1" y="yy1" />
</foo>
```

```
<foo>
  <bar x="xx1" y="yy1" />
  <bar x="xx2" y="yy2" />
</foo>
```

I want to get an output of the number of `bar` tags.  The resulting fight I had with yq is borderline insanity.  There are many factors that make it extremely difficult:

**Describe the solution you'd like**

This is kind of difficult, since there's many different ways in which this could work:

* Could have some flag that creates a consistent type: so parsing either of the above xmls would return the same structure (ie, dropping the "convenience" of not always using arrays).

* There is a `length`, but it is happy to work on any input, so piping anything to `length` is a losing battle AFAICT.  Could work if there was some `possibleArrayLength` for values that are possibly wrapped in an array, giving the length of the array or 1 if it wasn't wrapped.  (ie, dropping the "convenient" aspect of `length` being defined on any input.)

* I even tried `| type` and quickly discovered that there's no if-then-else that can depend on it.  The best I could get is pretty horrifying:

      ([.foo.bar | select(type=="!!map")]) + (.foo.bar | select(type=="!!seq"))

  and that still fails since in a single case the RHS is empty rather than `[]`.

**Comment by mikefarah** [comment]
If you know a particular path is an array (but sometimes just has one thing), you can use this trick to ensure it's always an array:

```
yq -oy '(.foo.bar) |= [] + .' file.xml
```

This will set `.foo.bar` to be an empty array concatenated with the original value. If originally it was empty, then its an empty array. If its single thing, then its an array with one thing. If it was an array, then it stays as an array :)

Hope that helps! I thought I had that as an example in the docs, but it seems I don't (sorry). Will add this is an example 👍🏼

**Comment by ebarzilay-godaddy** [comment]
@mikefarah Thanks!  That indeed works, I don't know how I missed it...

Just to make it into a more complete thing, the magic expression that IMO would be very useful to show specifically when talking about parsing xml is:

    [] + .blah // []

which returns a list consistently, even when there are none.

(I'd close it, but maybe you want to keep it open as a reminder...)
